blockstack/gaia
# Configure maximum concurrent upstreams

## Goal of PR

When a hub is deployed as a multi-user service, and receiving many concurrent requests, it can eventually exhaust OS sockets with too many concurrent upstream sockets (the nodejs "http agent pool" will grow with asynchronous request.)

By default, the current behavior remains. A setting of something like 100 should be sufficient to deal with that exhaustion issue without substantially reducing the real throughput of the server.

## Implementation

This simply sets the `[https|http].globalAgent.maxSockets` setting from a config variable (which can also be set via an env var: `GAIA_MAX_UPSTREAM`

## Manual Testing

To test this, spawn a hub server without the config variable set:

```bash
$ CONFIG_PATH=$PWD/sample-configs/config.aws.json npm run start
```

Then, in another terminal, issue 10 or so concurrent POSTs:

```bash
 $ time node gaiaTest.js 
[DEBUG] connectToGaiaHub: http://localhost:5000/hub_info
[DEBUG] uploadToGaiaHub: uploading foo0.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo1.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo2.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo3.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo4.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo5.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo6.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo7.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo8.txt to http://localhost:5000
[ 'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo0.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo1.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo2.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo3.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo4.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo5.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo6.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo7.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/19zS3btKKTz3RXbVNPDBwgMi1f7ZHusSyT/foo8.txt' ]

real    0m0.784s
user    0m0.493s
sys     0m0.033s
```

You can see the requests are being issued concurrently.

Now, stop the hub server, and restart it with `GAIA_MAX_UPSTREAM` set to 1:

```bash
$ GAIA_MAX_UPSTREAM=1 CONFIG_PATH=$PWD/sample-configs/config.aws.json npm run start
```

```bash
$ time node gaiaTest.js 
[DEBUG] connectToGaiaHub: http://localhost:5000/hub_info
[DEBUG] uploadToGaiaHub: uploading foo0.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo1.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo2.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo3.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo4.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo5.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo6.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo7.txt to http://localhost:5000
[DEBUG] uploadToGaiaHub: uploading foo8.txt to http://localhost:5000
[ 'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo0.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo1.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo2.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo3.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo4.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo5.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo6.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo7.txt',
  'https://gaia-test-driver-bucket.s3.amazonaws.com/1EWxH7KAT2193Zvj2WsJNrB8hs7pK8G543/foo8.txt' ]

real    0m1.276s
user    0m0.468s
sys     0m0.028s
```

You can see that there's no asynchrony on the upstreams on the hub this time.

## Automated Testing

There's no automated testing here, since testing would probably require timing that would probably false fail pretty frequently.

## Submission Checklist


      contains correct TypeFlow annotations.


      would be expected to interact with them.
